cache/cache.go

```diff
@@ -289,3 +303,93 @@ func (p2 *progressWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
 	p2.task.Add(n)
 	return n, errors.WithStack(err)
 }
+
+// matches: https://github.com/{OWNER}/{REPO}/releases/download/{TAG}/{ARTEFACT}
+var githubRe = regexp.MustCompile(`^https\://github.com/([^/]+)/([^/]+)/releases/download/([^/]+)/([^/]+)$`)
+
+type githubReleaseInfo struct {
+	owner, repo, tag, asset, token string
+}
+
+func getGithubReleaseInfo(uri string) (*githubReleaseInfo, bool) {
+	m := githubRe.FindStringSubmatch(uri)
+	if len(m) != 5 {
+		return nil, false
+	}
+	token := os.Getenv("HERMIT_GITHUB_TOKEN")
+	if token == "" {
+		return nil, false
+	}
+	g := &githubReleaseInfo{token: token}
+	var err error
+	if g.owner, err = url.PathUnescape(m[1]); err != nil {
+		return nil, false
+	}
+	if g.repo, err = url.PathUnescape(m[2]); err != nil {
+		return nil, false
+	}
+	if g.tag, err = url.PathUnescape(m[3]); err != nil {
+		return nil, false
+	}
+	if g.asset, err = url.PathUnescape(m[4]); err != nil {
+		return nil, false
+	}
+	return g, true
+}
+
+// GHRelease is a minimal type for GitHub releases meta information
+// retrieved via the GitHub API see:
+// https://docs.github.com/en/rest/reference/repos#list-releases
+type GHRelease struct {
+	TagName string    `json:"tag_name"`
+	Assets  []GHAsset `json:"assets"`
+}
+
+// GHAsset is a minimal type for assets in the GitHub releases
+// meta information retrieved via the GitHub API see:
+// https://docs.github.com/en/rest/reference/repos#list-releases
+type GHAsset struct {
+	Name string `json:"name"`
+	ID   int    `json:"id"`
+}
+
+func downloadGHPrivate(client *http.Client, ghi *githubReleaseInfo, file string) (w *os.File, response *http.Response, err error) {
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	header := http.Header{"Authorization": []string{"token " + ghi.token}}
+	u := fmt.Sprintf("https://api.github.com/repos/%s/%s/releases", ghi.owner, ghi.repo)
+	req, err := http.NewRequest("GET", u, nil) // nolint: noctx
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "cannot get private github repo assets")
+	}
+	req.Header = header.Clone()
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "cannot GET private github repo assets via API")
+	}
+	defer resp.Body.Close()
+	r := []GHRelease{}
+	if err := json.NewDecoder(resp.Body).Decode(&r); err != nil {
+		return nil, nil, errors.Wrap(err, "cannot decode private github repo assets")
+	}
+	assetID, err := getAssetID(r, ghi.tag, ghi.asset)
+	header.Set("Accept", "application/octet-stream")
+	u = fmt.Sprintf("https://api.github.com/repos/%s/%s/releases/assets/%d", ghi.owner, ghi.repo, assetID)
+	return Download(client, header, u, file)
+}
+
+func getAssetID(releases []GHRelease, tag, assetName string) (int, error) {
+	for _, r := range releases {
+		if r.TagName != tag {
+			continue
+		}
+		for _, a := range r.Assets {
+			if a.Name == assetName {
+				return a.ID, nil
+			}
+		}
+	}
+	return -1, fmt.Errorf("cannot find asset %s %s", tag, assetName)
```

Comment:
I think you should return the `int` zero value (0) here for the error case. It is meaningless in the error case but setting it to anything other than the zero value implies it has meaning.

Reply:
changed to string and url.